tests/interlocked_int/interlocked_int.c

```diff
@@ -0,0 +1,855 @@
+//Copyright(c) Microsoft.All rights reserved.
+//Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifdef __cplusplus
+#include <cstdlib>
+#include <cstddef>
+#include <cstdint>
+#else
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdint.h>
+#endif
+
+
+#include "testrunnerswitcher.h"
+
+static TEST_MUTEX_HANDLE g_testByTest;
+
+
+
+#include "azure_c_util/interlocked.h"
+
+
+BEGIN_TEST_SUITE(interlocked_int)
+
+
+TEST_SUITE_INITIALIZE(a)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+
+    
+}
+
+TEST_SUITE_CLEANUP(b)
+{
+
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(c)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+
+}
+
+TEST_FUNCTION_CLEANUP(d)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_001 [ interlocked_add shall atomically add *addend with value and store the result in *addend.]
+Tests_SRS_INTERLOCKED_43_032: [interlocked_add shall return the result of the addition.]
+*/
+
+TEST_FUNCTION(interlocked_add_does_addition)
+{
+    ///arrange
+    volatile int32_t addend = INT32_MAX;
+    int32_t value = INT32_MIN;
+
+    ///act
+    int32_t return_val = interlocked_add(&addend, value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(int32_t, -1, addend);
+    ASSERT_ARE_EQUAL(int32_t, -1, return_val);
+}
+
+/*
+Tests_SRS_INTERLOCKED_43_002[interlocked_and shall perform an atomic bitwise AND operation on the 32 - bit integer values * destination and value and store the result in * destination.]
+Tests_SRS_INTERLOCKED_43_033: [interlocked_and shall return the initial value of * destination.]
+*/
+
+TEST_FUNCTION(interlocked_and_does_bitwise_and)
+{
+    ///arrange
+    volatile uint32_t destination = 0xF0F0F0F0;
+    uint32_t value = 0x0F0F0F0F;
+
+    ///act
+    uint32_t return_val = (uint32_t)interlocked_and((volatile int32_t*)&destination, (int32_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(uint32_t, 0, destination);
```

Comment:
>0 [](start = 31, length = 1)

I'd consider changing the values so this results in a non-zero value. 0 seems like a default value, so the test might be a little better to check for something else (e.g. what if the implementation was just setting destination to 0 every time?).

Reply:
This is for here and everywhere

---
In reply to: [427689648](https://github.com/Azure/azure-c-util/pull/43#discussion_r427689648) [](ancestors = 427689648)